updates internal framework for python functions

ohasselblad opened the pull request:
@rochoa, could you take a look at this?

This branch updates the framework for accessing the Python functions in this extension. I stripped out the non-spatial kmeans function and will open a new feature branch for this function.

ohasselblad commented:
@talos and @stuartlynn, could you give the file `src/py/crankshaft/crankshaft/clustering/kmeans.py` a code review here? There are some outstanding issues with writing unit tests, so I haven't added those to a commit yet (namely, it looks like I would have to add an additional class to mockplpy if I use the methods of the database response object like `db_response.colnames()`).

stuartlynn commented on src/pg/sql/11_kmeans.sql:
```diff
@@ -1,21 +1,40 @@
+  id_col TEXT DEFAULT 'cartodb_id',
+  standarize BOOLEAN DEFAULT true
+)
+RETURNS TABLE(cluster_label text, cluster_center text, rowid bigint) AS $$
```
I am wondering if we should return the cluster center as a json object with keys for the variable name and its mean value in the cluster rather than text? Not sure how easy it is to return json from plpython

talos replied:
Just to but in -- even if JSON isn't properly supported, it should be straightforward to return stringified JSON from Plpython, then cast it back to JSON using a PL/PGSQL wrapper of the PL/Python function.

ohasselblad replied:
I was thinking of stringifying it anyway since the data type mappings between sql and plpython are pretty limited, so will be doing what you suggest @talos

ohasselblad replied:
Done, but the output in carto isn't optimal for display:
<img width="326" alt="screen shot 2016-11-10 at 11 47 26" src="https://cloud.githubusercontent.com/assets/1041056/20185883/d3c136f0-a73b-11e6-8832-2c9b9724e825.png">

Outputting as a string will allow people to view the data in the builder, and my guess is that if people need to access it they may have the chops to know how to convert to a postgres json object and then parse it... On the other hand, we could ask for better display of json in the builder data view.

stuartlynn commented on src/py/crankshaft/crankshaft/clustering/kmeans.py:
```diff
@@ -1,18 +1,98 @@
+    return zip(ids, labels)
+
+
+def kmeans_nonspatial(query, colnames, num_clusters=5,
```
It might also be worthwhile calculating a [Silhouette](https://en.wikipedia.org/wiki/Silhouette_(clustering)) score for each cluster and returning that as well. This would give us a metric of how good each cluster is. You can do this relatively easily in sklearn : http://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html

ohasselblad replied:
👍 done :)

ohasselblad replied:
I actually did `metrics.silhouette_sample` http://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_samples.html

rochoa commented on src/py/crankshaft/crankshaft/query_runner.py:
```diff
@@ -0,0 +1,43 @@
+import plpy
+
+
+class QueryRunner:
```
As this is attached to moran/markov queries, I would go with a name closer to that. QueryRunner is quite generic, while get_markov is very specific. What about something like `AnalysisDataProvider`?

ohasselblad replied:
👍 I like `AnalysisDataProvider`

rochoa commented on src/py/crankshaft/crankshaft/query_runner.py:
```diff
@@ -0,0 +1,43 @@
+
+
+class QueryRunner:
+    def get_markov(self, query):
```
Why do you need to provide the query? I would try to encapsulate the query generation within this class, making this method to handle the required params (num_ngbrs,...).

ohasselblad replied:
Ok, makes sense. And `__init__` would construct the query.. Which would clean up the code in other functions like `Moran.local_stat` :)

ohasselblad replied:
Done

rochoa commented on src/py/crankshaft/crankshaft/query_runner.py:
```diff
@@ -0,0 +1,43 @@
+        except plpy.SPIError, err:
+            plpy.error('Analysis failed: %s' % err)
+
+    def get_moran(self, query):
```
Similar approach in all the public functions in this class.

ohasselblad replied:
done.

ohasselblad commented:
Ok @rochoa, made those changes. What do you think?

rochoa commented:
Sorry for the delay @ohasselblad.

It looks better for me 👍 .

ohasselblad commented:
@ethervoid I updated one function from a new merge that didn't have the new structure yet, and removed `plpy` imports that are no longer needed. Looks like there's another failure for the postgres dependency (`postgresql-9.5=9.5.2-3cdb2`). Otherwise, looks good!
